File: import/download.js
const fs = require('fs');
const path = require('path');
const { spawn, execSync } = require('child_process');
const https = require('https');

const IMPORT_DIR = __dirname;
const LINKS_FILE = path.join(IMPORT_DIR, 'links.txt');
const PLAYLIST_DIR = path.join(IMPORT_DIR, 'playlist');
const BIN_DIR = path.join(IMPORT_DIR, 'bin');
const YTDLP_PATH = path.join(BIN_DIR, 'yt-dlp');

// Create directories if they don't exist
if (!fs.existsSync(PLAYLIST_DIR)) fs.mkdirSync(PLAYLIST_DIR, { recursive: true });
if (!fs.existsSync(BIN_DIR)) fs.mkdirSync(BIN_DIR, { recursive: true });

// Create links.txt if it doesn't exist
if (!fs.existsSync(LINKS_FILE)) {
  fs.writeFileSync(LINKS_FILE, `# Paste your YouTube video or playlist URLs here (one per line)
# Lines starting with # are ignored.
https://youtube.com/playlist?list=PL-Vr7Z_y18M7k9nuVyw3qsYU93XMCby1r&si=R-SRas9JBiOp7TzJ
`, 'utf8');
  console.log(`Created ${LINKS_FILE}. Please paste your YouTube URLs inside this file.`);
}

// Download yt-dlp binary if it doesn't exist
async function ensureYtdlp() {
  if (fs.existsSync(YTDLP_PATH)) {
    const stats = fs.statSync(YTDLP_PATH);
    if (stats.size > 0) {
      return YTDLP_PATH;
    }
    fs.unlinkSync(YTDLP_PATH);
  }

  console.log('Downloading standalone yt-dlp binary for macOS...');
  const url = 'https://github.com/yt-dlp/yt-dlp/releases/latest/download/yt-dlp';

  try {
    execSync(`curl -L -o "${YTDLP_PATH}" "${url}"`);
    fs.chmodSync(YTDLP_PATH, '755'); // Make executable
    console.log('yt-dlp downloaded and configured successfully.');
    return YTDLP_PATH;
  } catch (err) {
    console.error('Failed to download yt-dlp via curl', err);
    throw err;
  }
}

function getUrls() {
  const content = fs.readFileSync(LINKS_FILE, 'utf8');
  return content
    .split('\n')
    .map(line => line.trim())
    .filter(line => line && !line.startsWith('#'));
}

async function downloadSong(url, ytdlpPath) {
  console.log(`\n--------------------------------------------`);
  console.log(`Processing: ${url}`);

  return new Promise((resolve) => {
    // Format selector: bestaudio of extension m4a (aac) to avoid transcoding dependencies (ffmpeg)
    const args = [
      '--no-check-certificate', // Bypass macOS Python SSL verification issues
      '-f', 'ba[ext=m4a]/ba', // Get best audio, prioritizing native m4a
      '-o', path.join(PLAYLIST_DIR, '%(title)s.%(ext)s'),
      url
    ];

    const child = spawn('python3', [ytdlpPath, ...args]);

    child.stdout.on('data', (data) => {
      process.stdout.write(data.toString());
    });

    child.stderr.on('data', (data) => {
      process.stderr.write(data.toString());
    });

    child.on('close', (code) => {
      if (code === 0) {
        console.log(`Successfully downloaded.`);
        resolve(true);
      } else {
        console.error(`Download failed with exit code: ${code}`);
        resolve(false);
      }
    });
  });
}

async function run() {
  try {
    const ytdlpPath = await ensureYtdlp();
    const urls = getUrls();

    if (urls.length === 0) {
      console.log('No URLs found in links.txt. Add some links and run the script again.');
      return;
    }

    console.log(`Found ${urls.length} URLs to download...`);
    for (const url of urls) {
      await downloadSong(url, ytdlpPath);
    }
    console.log('\nAll downloads complete! Files saved in: import/playlist/');
  } catch (error) {
    console.error('An error occurred during execution:', error);
  }
}

run();

File: import/download.py
import os
import subprocess
import sys

IMPORT_DIR = os.path.dirname(os.path.abspath(__file__))
LINKS_FILE = os.path.join(IMPORT_DIR, 'links.txt')
PLAYLIST_DIR = os.path.join(IMPORT_DIR, 'playlist')
BIN_DIR = os.path.join(IMPORT_DIR, 'bin')
YTDLP_PATH = os.path.join(BIN_DIR, 'yt-dlp')

def get_urls():
    if not os.path.exists(LINKS_FILE):
        print(f"Links file not found at {LINKS_FILE}")
        return []
    
    urls = []
    with open(LINKS_FILE, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith('#'):
                urls.append(line)
    return urls

def download_song(url):
    print("\n" + "-"*44)
    print(f"Processing: {url}")
    
    # Arguments identical to the node script:
    # -f ba[ext=m4a]/ba -o playlist/%(title)s.%(ext)s
    output_template = os.path.join(PLAYLIST_DIR, '%(title)s.%(ext)s')
    cmd = [
        sys.executable,
        YTDLP_PATH,
        '--no-check-certificate',
        '-f', 'ba[ext=m4a]/ba',
        '-o', output_template,
        url
    ]
    
    try:
        # Run and stream stdout/stderr to the console
        result = subprocess.run(cmd, check=True)
        if result.returncode == 0:
            print("Successfully downloaded.")
            return True
    except subprocess.CalledProcessError as e:
        print(f"Download failed with exit code: {e.returncode}")
        return False
    except Exception as e:
        print(f"An error occurred: {e}")
        return False

def main():
    if not os.path.exists(PLAYLIST_DIR):
        os.makedirs(PLAYLIST_DIR, exist_ok=True)
        
    urls = get_urls()
    if not urls:
        print("No URLs found in links.txt. Add some links and run the script again.")
        return
        
    print(f"Found {len(urls)} URLs to download...")
    success_count = 0
    for url in urls:
        if download_song(url):
            success_count += 1
            
    print(f"\nAll downloads complete! {success_count}/{len(urls)} songs downloaded.")
    print(f"Files saved in: {PLAYLIST_DIR}")

if __name__ == '__main__':
    main()
